pkg/reconciler/pipeline/pipeline.go

```diff
@@ -0,0 +1,289 @@
+	}
+	p.Status.PropagateSubscriptionStatuses(subs)
+
+	return nil
+}
+
+func (r *Reconciler) updateStatus(ctx context.Context, desired *v1alpha1.Pipeline) (*v1alpha1.Pipeline, error) {
+	p, err := r.pipelineLister.Pipelines(desired.Namespace).Get(desired.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	// If there's nothing to update, just return.
+	if reflect.DeepEqual(p.Status, desired.Status) {
+		return p, nil
+	}
+
+	// Don't modify the informers copy.
+	existing := p.DeepCopy()
+	existing.Status = desired.Status
+
+	return r.EventingClientSet.MessagingV1alpha1().Pipelines(desired.Namespace).UpdateStatus(existing)
+}
+
+func (r *Reconciler) reconcileChannel(ctx context.Context, channelName string, channelResourceInterface dynamic.ResourceInterface, p *v1alpha1.Pipeline) (*unstructured.Unstructured, error) {
+	c, err := channelResourceInterface.Get(channelName, metav1.GetOptions{})
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			newChannel, err := resources.NewChannel(channelName, p)
+			logging.FromContext(ctx).Error(fmt.Sprintf("Creating Channel Object: %+v", newChannel))
+			if err != nil {
+				logging.FromContext(ctx).Error(fmt.Sprintf("Failed to create Channel resource object: %s/%s", p.Namespace, channelName), zap.Error(err))
+				return nil, err
+			}
+			created, err := channelResourceInterface.Create(newChannel, metav1.CreateOptions{})
+			if err != nil {
+				logging.FromContext(ctx).Error(fmt.Sprintf("Failed to create Channel: %s/%s", p.Namespace, channelName), zap.Error(err))
+				return nil, err
+			}
+			logging.FromContext(ctx).Error(fmt.Sprintf("Created Channel: %s/%s", p.Namespace, channelName), zap.Any("NewChannel", zap.Any("NewChannel", newChannel)))
+			return created, nil
+		} else {
+			logging.FromContext(ctx).Error(fmt.Sprintf("Failed to get Channel: %s/%s", p.Namespace, channelName), zap.Error(err))
+			return nil, err
+		}
+	}
+	logging.FromContext(ctx).Error(fmt.Sprintf("Found Channel: %s/%s", p.Namespace, channelName), zap.Any("NewChannel", c))
+	return c, nil
+}
+
+func (r *Reconciler) reconcileSubscription(ctx context.Context, step int, p *v1alpha1.Pipeline) (*eventingv1alpha1.Subscription, error) {
+	expected := resources.NewSubscription(step, p)
+
+	subName := resources.PipelineSubscriptionName(p.Name, step)
+	sub, err := r.subscriptionLister.Subscriptions(p.Namespace).Get(subName)
+
+	// If the resource doesn't exist, we'll create it.
+	if apierrs.IsNotFound(err) {
+		sub = expected
+		logging.FromContext(ctx).Info(fmt.Sprintf("Creating subscription: %+v : SUBSCRIBERSPEC: %+v", sub, sub.Spec.Subscriber))
+		newSub, err := r.EventingClientSet.EventingV1alpha1().Subscriptions(sub.Namespace).Create(sub)
+		if err != nil {
+			// TODO: Send events here, or elsewhere?
+			//			r.Recorder.Eventf(p, corev1.EventTypeWarning, subscriptionCreateFailed, "Create Pipeline's subscription failed: %v", err)
+			return nil, err
```

Comment:
The error should be logged

Reply:
it gets logged if this method fails at the call site on L208, seemed redundant to log it twice?

Reply:
added a log statement regardless.

Reply:
My thoughts have been going the other way. Rather than log at each function level. Attach the additional context at each function level, then the log only when it is no longer 'fatal' (for reconcilers, normally the `reconcile()` method, which recovers from the error).

So this function would be something like:

```golang
if err != nil {
    return nil, fmt.Errorf("failed to create subscription object: %v", err)
}
```

Then 208 (now 217) becomes something like:

```golang
if err != nil {
    return nil, fmt.Errorf("failed to reconcile Subscription Object for step %d: %v", step, err)
}
```

Which ends up getting logged exactly once in the reconcile() method (line 137).